### PR TITLE
Consuming iterator

### LINKED ISSUE #4: Add a consuming in-order iterator

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #4.

### CODE REVIEW

**Comment by FlashCat** [comment]
r? @huonw

(I've picked a reviewer for you, use r? to override)

File: src/lib.rs [hunk 1]
```diff
@@ -919,6 +1018,32 @@ impl<'a, K: Hash + Eq, V, S: BuildHasher> IntoIterator for &'a mut LinkedHashMap
...
+            unsafe { ((*self.head).prev, (*self.head).next) }
+        } else {
+            (ptr::null_mut(), ptr::null_mut())
+        };
+        let len = self.len();
+
+        unsafe { drop_empty_entry_box(self.head) }
+        self.clear_free_list();
+        // drop the HashMap but not the LinkedHashMap
+        mem::replace(&mut self.map, unsafe { mem::uninitialized() });
```
Comment (apasel422) [inline src/lib.rs]: Can this just be

```rust
self.map = unsafe { mem::uninitialized() };
```

File: tests/test.rs [hunk 2]
```diff
@@ -178,6 +178,28 @@ fn test_iter_mut() {
 }
 
 #[test]
+fn test_consuming_iter() {
```
Comment (apasel422) [inline tests/test.rs]: Can you add a test case here for an empty map?

File: src/lib.rs (hunk 1 above)
Comment (dtolnay) [inline src/lib.rs]: Yes it can, good catch.

File: tests/test.rs (hunk 2 above)
Comment (dtolnay) [inline tests/test.rs]: Done. Also added one for a non-empty free list.

**Comment by apasel422** [comment]
Looks good. It might be beneficial to have one more test that ensures that the every value is dropped exactly once:

```rust
#[test]
fn test_into_iter_drop() {
    struct Counter<'a>(&'a mut usize);

    impl<'a> Drop for Counter<'a> {
        fn drop(&mut self) {
            *self.0 += 1;
        }
    }

    let mut a = 0;
    let mut b = 0;
    let mut c = 0;

    {
        let mut map = LinkedHashMap::new();
        map.insert("a", Counter(&mut a));
        map.insert("b", Counter(&mut b));
        map.insert("c", Counter(&mut c));

        let mut iter = map.into_iter();
        iter.next();
        iter.next_back();
    }

    assert_eq!(a, 1);
    assert_eq!(b, 1);
    assert_eq!(c, 1);
}
```

**Comment by dtolnay** [comment]
Nice. I added it.
